docs/webrocket.3.txt
webrocket(3)
============

NAME
----
webrocket - The core library for webrocket tools

SYNOPSIS
--------
	import "webrocket"
	
DESCRIPTION
-----------
The *webrocket*(3) is a base library for all other webrocket tools.
It provides functionality to build full stack application listening
on WebSocket and Backend endpoints. Check its documentation to get
more information and knowledge.

EXAMPLES
--------
A trivial example app:

-----
package main

import (
    "webrocket"
    "os/signal"
)
	
func main() {
    ctx := webrocket.NewContext()
    v, _ := ctx.AddVhost("/example")
    v.OpenChannel("example")	
	
    websocket := ctx.NewWebsocketEndpoint("", 8080)
    go websocket.ListenAndServe()
	
    backend := ctx.NewBackendEndpoint("", 8081)
    go backend.ListenAndServe()
	
    <-signal.Incoming
}
-----

SEE ALSO
--------
link:webrocket-server.1.html[*webrocket-server*(1)],
link:webrocket-ctl.1.html[*webrocket-ctl*(1)],
link:webrocket-monitor.1.html[*webrocket-monitor*(1)]

AUTHOR
------
Chris Kowalik <chris@nu7hat.ch>

COPYRIGHT
---------
Copyright (C) 2011 by Krzysztof Kowalik <chris@nu7hat.ch>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.


NEWS
Release history
===============

v0.3.0 (unreleased)
-------------------
* Improvements in WebSockets API (publish renamed to broadcast)
* Better errors handling
* Support for dynamically managed vhosts and users
* API for server remote controll
* Simplified 'webrocket-server' tool
* Added manual entries for the library and tools
* Better build tools

v0.2.0
------
* Refactored and standarized JSON-based protocol
* CLI adapted to new protocol
* Added more flexible configuration

v0.1.1
------
* Fixed panics caused by logging

v0.1.0
------
* Advanced logging (logging to files, separate loggers for hubs, ...)
* Fully-functional CLI with JSON config files support
* Optimized broadcasting
* Bugfixes

v0.0.1
------
* Default JSON handler
* Custom handlers interface
* Non-blocking broadcast
* Authentication support
* Simple CLI


webrocket.go
// Copyright 2011 Chris Kowalik (nu7hatch). All rights reserved.
// Use of this source code is governed by a BSD-style license that
// can be found in the LICENSE file.
//
// Package webrocket implements advanced WebSocket server with custom
// protocols support. 
package webrocket

import (
	"websocket"
	"http"
	"log"
	"os"
)

// handlerMap is a map of resource handlers.
type handlerMap map[string]Handler

// Server defines parameters for running an WebSocket server.
type Server struct {
	http.Server
	Log      *log.Logger
	handlers handlerMap
	certFile string
	keyFile  string
}

/*
Creates new rocket's server bound to specified addr.
A Trivial example server:

    package main

    import "webrocket"

    func main() {
         s := webrocket.NewServer("ws://localhost:8080")
         s.Handle("/echo", webrocket.NewJSONHandler())
         s.ListenAndServe()
    }
*/
func NewServer(addr string) *Server {
	s := new(Server)
	s.Addr, s.Handler = addr, http.NewServeMux()
	s.handlers = make(handlerMap)
	s.Log = log.New(os.Stderr, "S : ", log.LstdFlags)
	return s
}

/*
Registers payload handler under specified path. Handler have to implement
communication protocol callbacks.
*/
func (s *Server) Handle(path string, h Handler) {
	proxy, ok := h.Register(s, path)
	if ok == nil {
		s.Handler.(*http.ServeMux).Handle(path, proxy)
		s.handlers[path] = h
	}
}

/*
Listens on the TCP network address srv.Addr and handles requests on incoming
websocket connections.
*/
func (s *Server) ListenAndServe() os.Error {
	s.Log.Printf("About to listen on %s\n", s.Addr)
	err := s.Server.ListenAndServe()
	if err != nil {
		s.Log.Fatalf("Server startup error: %s\n", err.String())
	}
	return err
}

/*
Listens on the TCP network address srv.Addr and handles requests on incoming TLS
websocket connections.
*/
func (s *Server) ListenAndServeTLS(certFile, keyFile string) os.Error {
	s.Log.Printf("About to listen on %s", s.Addr)
	err := s.Server.ListenAndServeTLS(certFile, keyFile)
	if err != nil {
		s.Log.Fatalf("Secured server startup error: %s\n", err.String())
	}
	return err
}

// subscriberMap contains sockets subscribing given channel.
type subscriberMap map[*conn]int

/*
channel keeps information about specified channel and it's subscriptions.
It's hub is used to broadcast messages.
*/
type channel struct {
	name        string
	owner       *handler
	subscribers subscriberMap
	subscribe   chan subscription
	broadcast   chan broadcaster
}

// channelMap is a map of channels.
type channelMap map[string]*channel

// broadcaster is a function for spreading messages to all chanel's subscribers. 
type broadcaster func(ws *conn)

/*
subscription struct is used to modify channel subscription state
from within the handler.
*/
type subscription struct {
	reader *conn
	active bool
}

func newChannel(h *handler, name string) *channel {
	ch := &channel{name: name, owner: h, subscribers: make(subscriberMap)}
	ch.subscribe, ch.broadcast = make(chan subscription), make(chan broadcaster)
	go ch.hub()
	return ch
}

// Channel's hub manages subscriptions and broacdasts messages to all subscribers.
func (ch *channel) hub() {
	for {
		select {
		case s := <-ch.subscribe:
			ch.subscribers[s.reader] = 0, s.active
		case b := <-ch.broadcast:
			for reader := range ch.subscribers {
				b(reader)
			}
		}
	}
}

/*
Handler handlers all incoming requestes using defined protocol. Handler also
manages all registered channels.

Trivial custom handler:

    type MyHandler struct {
        channels map[string]
    }

    func (*h MyHandler) Register(id interface{}) {
        // initialize your handler here...
    }
*/
type Handler interface {
	Register(s *Server, id interface{}) (websocket.Handler, os.Error)
}